### PR TITLE
nixos/tests/cloud-init: fix the erroring out

### PR DESCRIPTION
**Pull request by vikanezrimaya**
###### Motivation for this change
The test was broken for a 1.5 months apparently? Well, now it passes. I wonder how this happened...

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kisik21** [comment]
cc @nlewo as maintainer

**Comment by nlewo** [comment]
@GrahamcOfBorg test cloud-init

**Review by nlewo — APPROVED** [review]

**Comment by nlewo** [comment]
@kisik21 the test is actually flaky. 
It timeouts on the CI and locally, sometimes it passes and sometimes it timeouts.
Could you try to run it several times on your machine (you can use `nix-build --check`)?

**Comment by kisik21** [comment]
I think I've noticed it hanging once before, but disregarded that. Thought it's a minor glitch.

What do we usually do with flaky tests? How could we stabilize them?

**Comment by nlewo** [comment]
We would need to understand why it is flaky and then fix it.
I would first try to reproduce interactively (`nix-build -A the-test.driver`). If it is not possible to reproduce interactively, I would then try to put some `sleep` during the test to see if it solves the issue. If it solves the issue, I would then identify which component needs to be waited and then find a way to know when this component is actually ready.

I didn't really look at the failure, but it could be because the cloud-init process is still initializing while the test is already running.

**Comment by kisik21** [comment]
@nlewo I think I caught it! It hangs on the line: `systemd[1]: Startup finished in xxxxs (kernel) + yyyys (userspace) = zm z's.` I wonder what's happening inside of the VM?

**Comment by kisik21** [comment]
Ok I grokked how to debug these with the interactive runners

**Comment by kisik21** [comment]
@nlewo apparently it's not flaky anymore, after I slapped a time.sleep in a place that felt right after some inspection. Tested at least 10 times in a row. Maybe around a 100 times, but my terminal window magically closed, trashing all the test results with it - but I saw it happening.

**Comment by nlewo** [comment]
Ok, that's a nice starting point: now, we know we have to wait for a service! Do you know which one (`cloud-init` or `sshd`)?
Because we would be much better to avoid a `sleep` in the test (it consumes unneeded seconds). Instead, we should find something to detect the service is ready.

**Comment by kisik21** [comment]
Maybe we should wait for multi-user.target instead of sshd? I'll try that once I checkout that branch again locally :3

**Comment by kisik21** [comment]
Initial testing results are promising, it doesn't fail.

**Comment by kisik21** [comment]
Hundred iterations and it didn't fail once. I think we're clear

**Comment by nlewo** [comment]
@GrahamcOfBorg test cloud-init

**Comment by nlewo** [comment]
@kisik21 I actually think waiting for `multi-user.target` works by chance. I think the issue is that  cloud-init is writing sshd keys while the ssh connection is creating.
I added a statement ensuring cloud-init terminates its run before continuing.
I run it several time on my laptop and I have not been able to reproduce the issue.

Could you try on your machine as well?

@GrahamcOfBorg test cloud-init

**Comment by kisik21** [comment]
I launched it. Will take a while.

**Comment by kisik21** [comment]
@nlewo It works, no timeouts on my machine. I squashed the commits and credited you as a co-author.

File: nixos/tests/cloud-init.nix
Comment (flokli) [inline nixos/tests/cloud-init.nix:54]: why the `s/machine/unnamed/`?

File: nixos/tests/cloud-init.nix
Comment (flokli) [inline nixos/tests/cloud-init.nix:57]: This is a bit hacky. I'd really prefer if we could wait for the the service that's supposed to be creating this. Is this part of `cloud-init.service` (On which we already wait in L54), or what other unit handles the `write_files` part?

Maybe we just need to wait on another one of the services? `cloud-final.service` maybe, or just the entire `cloud-config.target`?

File: nixos/tests/cloud-init.nix
Comment (kisik21) [reply nixos/tests/cloud-init.nix:54]: Because otherwise it wouldn't work! The test script will just crash on the first line. It's because the objects for test machines in the namespace are generated from system closure hostnames.

File: nixos/tests/cloud-init.nix
Comment (kisik21) [reply nixos/tests/cloud-init.nix:57]: This could work, but I'm not a cloud-init pro. I just wanted to fix a crashing test script and maaaaybe learn a bit of cloud-init?

File: nixos/tests/cloud-init.nix
Comment (flokli) [reply nixos/tests/cloud-init.nix:54]: I was used to these being called `machine`, and the tests working, so mostly confused myself here. Seems like this is now needs to be called `unnamed` here, where `machine` still works in other places, like `nixos/tests/sudo.nix` :shrug

File: nixos/tests/cloud-init.nix
Comment (flokli) [reply nixos/tests/cloud-init.nix:57]: If you want, you can take a look at `nixos/modules/services/system/cloud-init.nix`, and how we invoke with various parameters in different units, then see where the `write_files` modules is handled inside cloud-init.

It's probably easier to just wait on `cloud-config.target`, which should include all of them (I think)?

File: nixos/tests/cloud-init.nix
Comment (kisik21) [reply nixos/tests/cloud-init.nix:54]: It's probably because the hostname is unset now, and now the closure, instead of having a hostname after its name in the attribute set, now has a hostname equal to an empty string. NixOS handles this by tagging it as "unnamed".

File: nixos/tests/cloud-init.nix
Comment (kisik21) [reply nixos/tests/cloud-init.nix:57]: It looks like cloud-config.target includes write_files. I'm launching 200 iterations of the unit test with timeout of 120 seconds (it usually completes in 70 seconds on my machine) to test it. I guess I'll report results the next morning!

Comment (kisik21) [reply nixos/tests/cloud-init.nix:57]: Oh no, I've miscalculated the timeout? it timed out..

Comment (kisik21) [reply nixos/tests/cloud-init.nix:57]: Oh no, it looks like the test became more flaky. I guess I'll have to order after cloud-final - it includes things not included in cloud-config.target!

Comment (kisik21) [reply nixos/tests/cloud-init.nix:57]: Ok, after ordering the test after `cloud-final.service` it looks like everything's fine. I'll leave it to run overnight to see if there are any failures. If they are, I shall be greeted with a huge error on my console.

Comment (nlewo) [reply nixos/tests/cloud-init.nix:57]: Yep, would be better to be able to wait on an unit.

**Comment by nlewo** [comment]
Oh, and i actually learnt that for oneshot services, systemd waits for the process termination before continuing! That could explain why it was working when you were waiting `multi-user.target`!

**Comment by kisik21** [comment]
It also works when waiting for `cloud-final.service`. But now I'm confused as to who to credit in `Co-Authored-By` trailers: you both or one of you? i wanna do the most sensible thing but am confused :smile:

**Comment by flokli** [comment]
This is fine :-D Thanks a lot for fixing it!

**Pull request merged**
